### PR TITLE
🍰 First Implementation Of Filtering Posts By Language

### LINKED ISSUE #2045: 🚀 [Feature] Filter Posts by language
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-25T18:35:18Z" title="Friday, October 25th 2019, 8:35:18 pm +02:00">Oct 25, 2019</time>_
_Closed <time datetime="2019-11-04T09:20:14Z" title="Monday, November 4th 2019, 10:20:14 am +01:00">Nov 4, 2019</time>_
---

## :rocket: Feature

grafik

I think that entering HC as non German speaker may cause the impression of a pure German Network, because until now nearly all the posts you see are in German.
To avoid this, it should be possible to filter the post by language, and may be set the default of the filter to the selected language.

### Design & Layout
As indicated in the screenshot. May be it is better placed below the emotes.

### Additional context
Thinking about the possiibilty to save filters.

### PR DESCRIPTION
**Pull request by Tirokk**
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-28T13:21:01Z" title="Monday, October 28th 2019, 2:21:01 pm +01:00">Oct 28, 2019</time>_
_Merged <time datetime="2019-11-04T09:20:14Z" title="Monday, November 4th 2019, 10:20:14 am +01:00">Nov 4, 2019</time>_
---

## 🍰 Pullrequest
grafik

First implementation.

### Issues

- fixes #2045

### Todo

- [ ] Finding a way to display flags of the languages. See comments of #2045 .
- [ ] Add tests.

### CODE REVIEW

**Comment by Tirokk** [comment]
> [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-28T14:01:51Z" title="Monday, October 28th 2019, 3:01:51 pm +01:00">Oct 28, 2019</time>_

---


Hey @Mogge !!! Great work! 🚀💫
Thanks a lot for your contribution. 👍🏼

That looks and functions very good in my opinion @Mogge . Even if some things are still missing.
Tests for example. I added a checkbox to the PRs description.

I was first not aware that I have to scroll down in the filters menu to see the buttons.

And it is kind of discussable if this is at least the right place to have it. Or should it be a general setting of the users profile? @datenbrei

File: webapp/locales/en.json
Comment (Tirokk) [inline webapp/locales/en.json:92]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-28T14:07:29Z" title="Monday, October 28th 2019, 3:07:29 pm +01:00">Oct 28, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -87,7 +87,11 @@
     },
     "followers": {
       "label": "Users I follow"
-    }
+    },
+      "language": {
```


Please format with correct indent. ☝🏼

File: webapp/locales/de.json
Comment (Tirokk) [inline webapp/locales/de.json:91]: > [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-28T14:07:38Z" title="Monday, October 28th 2019, 3:07:38 pm +01:00">Oct 28, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -86,7 +86,11 @@
     },
     "followers": {
       "label": "Benutzern, denen ich folge"
-    }
+    },
+      "language": {
```


Please format with correct indent.

**Comment by Tirokk** [comment]
> [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-28T14:09:52Z" title="Monday, October 28th 2019, 3:09:52 pm +01:00">Oct 28, 2019</time>_

---


You have some build errors @Mogge . 
Please click on details and then on `The Build`:
See https://travis-ci.com/Human-Connection/Human-Connection/builds/133810500
👇🏼

**Comment by Tirokk** [comment]
> [Tirokk](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-28T14:11:22Z" title="Monday, October 28th 2019, 3:11:22 pm +01:00">Oct 28, 2019</time>_

---


And the flags are not absolutely related to the flags of the countries as well. @Mogge @datenbrei 
English language should have the English or the USA flag as an example … ?

Additional we need related countries on the posts and be able to filter for them as well. See #782

**Comment by Tirokk** [comment]
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-28T14:23:37Z" title="Monday, October 28th 2019, 3:23:37 pm +01:00">Oct 28, 2019</time>_

---


The problem with the flags is even greater in Spanish. I do not like flags at all. But it would be nice to have a graphical representation for each language, better than the flag icon I use now.
I will check the errors and try to write some tests.
I think it is more important to filter by language than by country by now. We want to have HC attractive for non-German speakers. When they enter the web and only see German post they may think it is only a German community. So may be by default we should show only posts in the chosen language to new users.

**Comment by Tirokk** [comment]
> [datenbrei](https://github.com/datenbrei) **Authored by [datenbrei](https://github.com/datenbrei)**
_<time datetime="2019-10-28T14:41:57Z" title="Monday, October 28th 2019, 3:41:57 pm +01:00">Oct 28, 2019</time>_

---


@Tirokk What would a country-specific post be? One about a country, one from a country? Related to a country? From a specific author of such a country? Or an author living in a country? 

First I would stick with just language, later first we have to decide, what we really want to do with it.

**Comment by Tirokk** [comment]
> [datenbrei](https://github.com/datenbrei) **Authored by [datenbrei](https://github.com/datenbrei)**
_<time datetime="2019-10-28T14:43:06Z" title="Monday, October 28th 2019, 3:43:06 pm +01:00">Oct 28, 2019</time>_

---


@Mogge and @Tirokk For now I'd take the english flag. It is the english language, spoken all over the world.

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-28T16:48:55Z" title="Monday, October 28th 2019, 5:48:55 pm +01:00">Oct 28, 2019</time>_

---


I think it's not important to worry about filtering by countries right now as well! This may very well be a feature we want to implement in the future, but certainly not in the short term. 

The reason we weren't using flags for the `LocaleSwitch` is for this very reason... I would say we should add the `Buttons` with the country code, like we do in `LocaleSwitch`... flags are prettier for sure, but there are many questions of which flag to show for which language, certainly the Brazil flag makes much more sense for `Portuguese`, no?

**Comment by Tirokk** [comment]
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-29T19:36:04Z" title="Tuesday, October 29th 2019, 8:36:04 pm +01:00">Oct 29, 2019</time>_

---


I pushed a new version with the tests added. When I ran yarn test I had no errors.

File: webapp/components/FilterPosts/FilterPosts.spec.js
Comment (Tirokk) [inline webapp/components/FilterPosts/FilterPosts.spec.js:114]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-31T10:15:52Z" title="Thursday, October 31st 2019, 11:15:52 am +01:00">Oct 31, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -97,13 +107,31 @@ describe('FilterPosts.vue', () => {
       expect(mutations['posts/TOGGLE_CATEGORY']).toHaveBeenCalledWith({}, 'cat4')
     })
 
+    it('calls TOGGLE_LANGUAGE when clicked', () => {
+      const wrapper = openFilterPosts()
+      englishButton = wrapper
+        .findAll('button.language-buttons')
+        .at(languages.findIndex(l => l.code === 'en'))
```


I really like your way of doing this! one thing I complain about with vue-test-utils over enzyme for example is that there is not `filterWhere`... but I really like how you've done this... I need to refactor some of my tests that look for the button in a position, which is easy to break if someone adds another button, for example.

File: webapp/components/FilterPosts/LanguageFilterMenuItems.vue
Comment (Tirokk) [inline webapp/components/FilterPosts/LanguageFilterMenuItems.vue:37]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-31T10:19:36Z" title="Thursday, October 31st 2019, 11:19:36 am +01:00">Oct 31, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,103 @@
+<template>
+  <ds-space margin-top="large">
+    <ds-flex id="filter-posts-header">
+      <ds-heading tag="h4">{{ $t('filter-posts.language.header') }}</ds-heading>
+      <ds-space margin-bottom="large" />
+    </ds-flex>
+    <ds-flex :gutter="{ lg: 'small' }">
+      <ds-flex-item
+        :width="{ base: '100%', sm: '100%', md: '100%', lg: '5%' }"
+        class="language-menu-item"
+      >
+        <ds-flex>
+          <ds-flex-item width="10%" />
+          <ds-flex-item width="100%">
+            <ds-button
+              icon="check"
+              @click.stop.prevent="resetLanguages"
+              :primary="!filteredLanguagesCodes.length"
+            />
+            <ds-flex-item>
+              <label class="language-labels">{{ $t('filter-posts.language.all') }}</label>
+            </ds-flex-item>
+            <ds-space />
+          </ds-flex-item>
+        </ds-flex>
+      </ds-flex-item>
+      <ds-flex-item :width="{ base: '0%', sm: '0%', md: '0%', lg: '4%' }" />
+      <ds-flex-item
+        :width="{ base: '0%', sm: '0%', md: '0%', lg: '3%' }"
+        id="languages-menu-divider"
+      />
+      <ds-flex v-for="language in locales" :key="language.code" class="languages-menu">
+        <ds-flex class="languages-menu">
+          <ds-flex-item width="100%" class="language-menu-item">
+            <ds-button
+              class="language-buttons"
+              icon="flag"
```


As I stated in the comments, I would prefer using the same country codes for these buttons as the `LocaleSwitch`, using flags of countries is problematic, as discussed, but also the same icon doesn't make a lot of sense to me either.

**Comment by Tirokk** [comment]
> [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-31T13:42:13Z" title="Thursday, October 31st 2019, 2:42:13 pm +01:00">Oct 31, 2019</time>_

---


grafik

As agreed with @mattwr18 I changed the flag icon to the language code.
A pinned post in a language that is not in the filter stays pinned and visible to the user.
I think about a solution for the language filter in the user profile page, as it is not supposed that a user changes the language filter a lot. The idea is adding a `languageFilter` property to the user, which is always applied when filtering posts.

File: webapp/components/FilterPosts/LanguageFilterMenuItems.vue
Comment (Tirokk) [reply webapp/components/FilterPosts/LanguageFilterMenuItems.vue:37]: > [Mogge](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-10-31T13:44:35Z" title="Thursday, October 31st 2019, 2:44:35 pm +01:00">Oct 31, 2019</time>_

---


grafik

**Comment by Tirokk** [comment]
> [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-31T16:30:59Z" title="Thursday, October 31st 2019, 5:30:59 pm +01:00">Oct 31, 2019</time>_

---


@mattwr18 are we good to go here?

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-31T21:45:31Z" title="Thursday, October 31st 2019, 10:45:31 pm +01:00">Oct 31, 2019</time>_

---


I would be ok to merge this, but if I am being picky @Mogge, I do ask myself why the buttons are rectangular when all the other filter buttons are circular... maybe I am being pedantic... what do you think @alina-beck?

Also, maybe we can remove the `[WIP]` from the title before we decide to merge or not @Mogge ??

**Comment by Tirokk** [comment]
> [alina-beck](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-11-04T06:55:24Z" title="Monday, November 4th 2019, 7:55:24 am +01:00">Nov 4, 2019</time>_

---


@mattwr18 @Mogge I thought the same when I saw this... We _are_ picky. :grin: But right now the style guide only allows `icons` for circular buttons so unless we want to save all language codes as svg icons I don't see that this can be solved. I am, however, planning to add this functionality when migrating the styleguide – and buttons will be one of the first things on my list! :)

**Review by Mogge — APPROVED** [review]
supercool

Wow, this is supercool!. For me it is already in a stage where we can merge it.

Regarding the flag discussion: I had the same talk with @appinteractive once and there is a whole blog why one shouldn't use flags to represent languages:
http://www.flagsarenotlanguages.com/blog/best-practice-for-presenting-languages/

Maybe we can add some more tests, e.g. for the store. We might want to increase the dropdown menu. Or just think how we can reduce this massive dropdown window. 

But to my eyes, this looks really great and already mergeable.

**Review by Mogge — APPROVED** [review]
Yup! @roschaefer @Mogge

**Pull request closed without merging**
